### PR TITLE
[internal/exp/metrics] Add functions to merge metrics

### PR DESCRIPTION
**Pull request by RichieSams**
**Description:**

This will merge the metrics in mdB into mdA, trying to re-use resourceMetrics, scopeMetrics, and metric values as possible. This will be used to help implement the new feature for: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/32513

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/32513 / https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/32690

**Testing:**

I created a unit test which tests various scenarios of how merge behavior should happen

**Documentation:**

The exported function is documented using standard golang style. And there are comments inside the code to explain what is going on and why

### CODE REVIEW

**Comment by RichieSams** [comment]
If someone can apply the `No Changelist` label, this should be good for final review and merge

File: internal/exp/metrics/streams/streams.go
Comment (fatsheep9146) [inline internal/exp/metrics/streams/streams.go:69]: The changes in this file seemed unrelated to this pr?

File: internal/exp/metrics/metrics.go
Comment (fatsheep9146) [inline internal/exp/metrics/metrics.go:18]: Could you add test case to cover this situation, `Any "unnecessary" duplicate entries in mdA will *not* be combined.`?

File: internal/exp/metrics/metrics.go
Comment (fatsheep9146) [inline internal/exp/metrics/metrics.go:86]: I suggest add `//exhaustive:enforce` lint check

File: internal/exp/metrics/testdata/basic_merge/b.yaml
Comment (fatsheep9146) [inline internal/exp/metrics/testdata/basic_merge/b.yaml:1]: Could test cover all types of metric, for example histogram?

File: internal/exp/metrics/metrics.go
Comment (jpkrohling) [inline internal/exp/metrics/metrics.go:15]: what are the benefits of merging it this way, instead of just adding mdB to a new resource metric on mdA? why would I want to use this, instead of just doing that? what are the perf implications? And why would this operation be here and not part of `pdata` itself?

File: internal/exp/metrics/metrics_test.go
Comment (jpkrohling) [inline internal/exp/metrics/metrics_test.go:17]: Can you add a benchmark as well? Perhaps one comparing a plain copyto from B into A vs. the new Merge function in this PR.

File: internal/exp/metrics/metrics.go
Comment (RichieSams) [reply internal/exp/metrics/metrics.go:86]: Oh cool. I didn't know you could do that. TIL

File: internal/exp/metrics/streams/streams.go
Comment (RichieSams) [reply internal/exp/metrics/streams/streams.go:69]: It's used in the datapoint merge function: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/32794/files#diff-756513e8ec6edd51e0dc85ef3e3756b9674a2929328395fc63e806e9d052da46R113

I'd also like to migrate `processor/intervalprocessor` to use this shared definition, instead of having a duplicate interface in multiple places. https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/intervalprocessor/internal/metrics/metrics.go

File: internal/exp/metrics/metrics.go
Comment (RichieSams) [reply internal/exp/metrics/metrics.go:15]: > what are the benefits of merging it this way, instead of just adding mdB to a new resource metric on mdA? why would I want to use this, instead of just doing that? 

To reduce data duplication during serialization. In the loadbalancingexporter, we split the incoming metric data by metric. Say our batch is 1 resource metric, 1 scope metric, and 4000 metrics. If we don't merge the resulting data, we end up with 4000 copies of the resource metric attributes and 4000 copies of the scope metric attributes. Which all has to be serialized to OTLP wire format.

By merging, we can reduce that back down to just 1 resource metric and 1 scope metric per group of metrics.

> What are the perf implications?

More CPU during processing time, to save CPU and bandwidth for the next step in the chain. It *should* be possible to measure it.

> And why would this operation be here and not part of pdata itself?
I assume you mean `pdatautil`? No good reason. If you feel it would be better there, then I can move it there instead.

File: internal/exp/metrics/metrics.go
Comment (RichieSams) [reply internal/exp/metrics/metrics.go:18]: Done

File: internal/exp/metrics/metrics_test.go
Comment (RichieSams) [reply internal/exp/metrics/metrics_test.go:21]: Done. Please double check my assumptions, but it looks to be similar performance (a copy is a copy), but depending on the situation, potentially drastically less memory allocations

```
Running tool: /usr/local/go/bin/go test -benchmem -run=^$ -bench . github.com/open-telemetry/opentelemetry-collector-contrib/internal/exp/metrics

goos: linux
goarch: amd64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/internal/exp/metrics
cpu: Intel(R) Xeon(R) CPU E5-2687W v4 @ 3.00GHz
BenchmarkMergeManyIntoSingle/Naive-24         	     100	  14482976 ns/op	 5990548 B/op	  130017 allocs/op
BenchmarkMergeManyIntoSingle/Deduplicating-24 	      79	  16174993 ns/op	 1671475 B/op	   30018 allocs/op
BenchmarkMergeManyIntoMany/Naive-24           	      96	  12998815 ns/op	 6106008 B/op	  130003 allocs/op
BenchmarkMergeManyIntoMany/Deduplicating-24   	     100	  14233288 ns/op	 1672314 B/op	   30019 allocs/op
BenchmarkNaiveMerge-24                        	1000000000	         0.0000004 ns/op	       0 B/op	       0 allocs/op
```

Comment (RichieSams) [reply internal/exp/metrics/metrics_test.go:21]: I could also potentially create a benchmark that compares serialization time with naive vs deduplicating merge

**Comment by RichieSams** [comment]
Any further questions or concerns?

**Comment by fatsheep9146** [comment]
@RichieSams could you fix the conflicts and failed lint checks?

File: internal/exp/metrics/metrics.go
Comment (jpkrohling) [reply internal/exp/metrics/metrics.go:15]: @open-telemetry/collector-approvers , I would need a second opinion here: would it be worth having this function as part of the pdata utilities, instead of having it here?

Comment (RichieSams) [reply internal/exp/metrics/metrics.go:15]: Bump

Comment (bogdandrutu) [reply internal/exp/metrics/metrics.go:15]: @jpkrohling in pdata you can merge without the need of a groupBy resource since that is faster. I don't think we can have a complicated logic like this directly in pdata, but if anything we can have a separate helper/util package with logic like this.

Comment (jpkrohling) [reply internal/exp/metrics/metrics.go:15]: Alright, in that case, I think it's OK to have it here in contrib as this PR proposes, and we can "promote" it to core if needed.

**Review by jpkrohling — APPROVED** [review]
It's OK to have this, especially as it's part of the internal API for this repo. Once we've seen a few places consuming this package, we can determine whether their usage deserves keeping it, or dealing with pdata directly.

**Comment by RichieSams** [comment]
Ok we're good to merge now. CI is green 🥳

**Pull request merged**
